File: lib/nn.dart
import 'gpu_tensor.dart';
import 'dart:math';

abstract class Module {
  List<Tensor> parameters();
  void zeroGrad() {
    for (var p in parameters()) p.zeroGrad();
  }

  void step(double lr) {
    for (var p in parameters()) p.step(lr);
  }
}

class Layer extends Module {
  late Tensor w;
  late Tensor b;
  final bool useGelu;

  Layer(int nin, int nout, {this.useGelu = true}) {
    // Xavier/Glorot Initialization
    double scale = sqrt(2.0 / (nin + nout));
    w = Tensor.fromList(
      [nin, nout],
      List.generate(nin * nout, (_) => (Random().nextDouble() * 2 - 1) * scale),
    );
    // Initialize bias slightly positive to prevent "dead" neurons
    b = Tensor.fill([1, nout], 0.01);
  }

  Tensor forward(Tensor x, List<Tensor> tracker) {
    // 1. xw = x * w
    final xw = x.matmul(w);

    // 2. out = xw + b (Broadcasting handled by our C++ max(shape) logic)
    final out = xw + b;

    tracker.addAll([xw, out]);

    if (useGelu) {
      // 3. act = gelu(out)
      final act = out.gelu();
      tracker.add(act);
      return act;
    }

    return out;
  }

  @override
  List<Tensor> parameters() => [w, b];
}

File: python/dart_cuda/core/layers/nn.py
"""Base Module + Layer, mirrors `lib/core/layers/nn.dart`."""

from __future__ import annotations

import math
import random
from abc import ABC, abstractmethod
from typing import List

from ..tensor.gpu_tensor import Tensor


class Module(ABC):
    @abstractmethod
    def parameters(self) -> List[Tensor]:
        ...

    def zero_grad(self) -> None:
        for p in self.parameters():
            p.zero_grad()

    zeroGrad = zero_grad

    def step(self, lr: float) -> None:
        for p in self.parameters():
            p.step(lr)


class Layer(Module):
    """Linear `nin -> nout` (+ optional GELU). Matches Dart `Layer`."""

    def __init__(self, nin: int, nout: int, *, useGelu: bool):
        scale = math.sqrt(2.0 / (nin + nout))
        rng = random.Random()
        w_vals = [(rng.random() * 2 - 1) * scale for _ in range(nin * nout)]
        self.w = Tensor.from_list([nin, nout], w_vals)
        self.b = Tensor.fill([1, nout], 0.01)
        self.use_gelu = useGelu

    # Dart-side name
    @property
    def useGelu(self) -> bool:
        return self.use_gelu

    def forward(self, x: Tensor, tracker: List[Tensor]) -> Tensor:
        xw = x.matmul(self.w)
        out = xw + self.b
        tracker.extend([xw, out])
        if self.use_gelu:
            act = out.gelu()
            tracker.append(act)
            return act
        return out

    def parameters(self) -> List[Tensor]:
        return [self.w, self.b]
